src/test/pybind/test_rados.py

```diff
@@ -825,6 +825,34 @@ def test_execute(self):
         ret, buf = self.ioctx.execute("foo", "hello", "say_hello", b"nose")
         eq(buf, b"Hello, nose!")
 
+    def test_aio_execute(self):
+        retval = [None]
+        lock = threading.Condition()
+        def cb(_, buf):
+            with lock:
+                retval[0] = buf
+                lock.notify()
+        self.ioctx.write("foo", b"") # ensure object exists
+
+        comp = self.ioctx.aio_exec("foo", "hello", "say_hello", b"", oncomplete=cb)
+        comp.wait_for_complete()
+        with lock:
+            while retval[0] is None:
+                lock.wait()
+        eq(comp.get_return_value(), 13)
+        eq(retval[0], b"Hello, world!")
+
+        retval[0] = None
+        comp = self.ioctx.aio_exec("foo", "hello", "say_hello", b"nose", oncomplete=cb)
```

Comment:
It's a bit of an historical artifact at this point - for reads, onsafe and oncomplete are both called at the same time by librados (reads only return on reply on the wire).

In the past, writes sent two replies, one when the data was readable (oncomplete) and one when the data was persisted (onsafe). These were only relevant for btrfs really, since xfs and other update-in-place filesystems ended up persisting and making data readable at the same time. Nice to be sure using the option doesn't do something silly and cause a crash though.